Product Vulnerability Process (#1103)
mattermost/mattermost-handbook: operations/security/product-security/product-vulnerability-process.md

# Product Vulnerability process

## Overview

> This document aims to provide guidance on the handling of product vulnerabilities found either internally or externally. 

Reviewer: ```suggestion
This document aims to provide guidance on the handling of product vulnerabilities. 
```

Nit, this bit seems unnecessary. That's just all the vulnerabilities, right? 🙂

It should explain to any of the relevant stakeholders all parts of the process and provide background information and 

> definitions for the daily usage.

Reviewer: ```suggestion
It describes the process and provides background information and 
definitions for the daily usage of all relevant stakeholders.
```

Nits:

 - `should` is a goal and belongs in the Goals section
 - As long as this isn't at 100%, it probably doesn't cover all parts down to the last detail
 - `any` → `all`

## Goals

> Establish and document a way of working that:

Reviewer: ```suggestion
Establish and document a way of working that ensures:
```

Nit, align wording with bullet points

> * Vulnerability are triaged efficiently and quick 

Reviewer: ```suggestion
* Vulnerabilities are triaged efficiently and quickly 
```

Nit, typos

> * Ownership and expectations for all parties are aligned

Reviewer: ```suggestion
* All parties are aligned around ownership and expectations
```

Nit, seems a bit clearer this way

> * Patch releases are available within the agreed SLAs

Reviewer: Do we have / plan to implement actual proper SLAs? Or would SLO (or some other related term) be more accurate? An internal SLA seems like it's really more of an SLO.

## Scope
This process covers any vulnerabilities found in officially published Mattermost products and that were either found 

> internally, or reported by a customer, bug bounty program or another third party.

Reviewer: ```suggestion
This process covers any vulnerabilities found in officially published Mattermost products that were either found 
internally, reported to us by a customer, or come in through the bug bounty program or via another third party.
```

Nits

> Infrastructure vulnerabilities are explicitly not covered as part of this process and owned by the Security Operations Team.

Reviewer: ```suggestion
Infrastructure vulnerabilities are explicitly out of scope for this process. Vulnerability management, as it relates to infrastructure, is owned by the Security Operations Team.
```

Nit, seems slightly clearer this way

Should we link to some SecOps docs here?

## Process Definitions

### Responsible Parties
- Product Security Team: Process Ownership; Vulnerability Owner
- Development Team: Vulnerability Remediation
- Release Manager: Release Planning
- Build Team: Build Process

### Vulnerability Sources
Vulnerabilities are reported from a variety of sources that are explained below.

#### Internal Finding
Vulnerability has been discovered by the internal security, development, or QA team. Additional possibility is finding 

> through internal security tooling such as OWASP DepCheck.

Reviewer: ```suggestion
through internal security automation such as Dependency-Track or CodeQL.
```

#### Responsible Disclosure

> Vulnerability has been reported through our responsible disclosure program.

Reviewer: ```suggestion
Vulnerability has been reported through our [responsible disclosure program](https://mattermost.com/security-vulnerability-report/).
```

#### Bug Bounty Program

> Vulnerability has been reported through our Bug Bounty program running on HackerOne.

Reviewer: ```suggestion
Vulnerability has been reported through our [Bug Bounty program running on HackerOne](https://hackerone.com/mattermost?type=team).
```

<table>
  <thead>
    <tr>
      <th style="text-align:left">Rating</th>
      <th style="text-align:left">CVSS Score</th>
    </tr>
  </thead>
  <tbody>
        <tr>
            <td>None</td>
            <td>0.0</td>
        </tr>
        <tr>
            <td>Low</td>
            <td>0.1 - 3.9</td>
        </tr>
        <tr>
            <td>Medium</td>
            <td>4.0 - 6.9</td>
        </tr>
        <tr>
            <td>High</td>
            <td>7.0 - 8.9</td>
        </tr>
        <tr>
            <td>Critical</td>
            <td>9.0 - 10.0</td>
        </tr>
  </tbody>
</table>

### Remediation SLAs

(INTERNAL - TBD to Publish)

### Issue Types
Vulnerabilities must be assigned one of the following vulnerability types:
- Authorization
- Reducing Attack Surface
- XSS
- Input Validation
- Phishing
- Injection
- Information Disclosure
- DOS
- Authentication
- CSRF
- Misconfiguration
- 3rd Party Library
- Remote Code Execution

If no fitting category can be found, the security engineer must request the inclusion of a new type from the product 
security team lead.

### Backport Policy

#### Vulnerabilities with Low Score to Medium Score:
- Upcoming Release

> - Extended Support Release

Reviewer 2: Let me know if this is a new policy change that I should follow. So far the policy has been the following (or what I was instructed to do when I took over release management a couple years ago):

 - Medium level vulnerabilities have the same backport policy as High/Critical level ones.
 - Low level vulnerabilities don't need to be backported.

#### Vulnerabilities with High to Critical Score:
- Upcoming Release
- Extended Support Release
- Last 3 Releases

## Process Steps

### Report
Vulnerability reports are received from internal or external parties via one of the methods outlined above. The Product 
Security Team must acknowledge to the reporter that it received the report within 24h on business days and will investigate 
whether it’s a valid issue. If the vulnerability was published on a news site/blog/etc. and is considered a public leak 

> the marketing team should be notified to coordinate on the response in the respective Mattermost channel with a “@channel” mention.

Reviewer: Which channel?

After the report was received, a “Security Vulnerability” ticket must be created in the Jira ticket that contains the 
following information:

* Title
* Description
* Category [If possible]
* Reporter
* Source

The security engineer that acknowledges the initial report must create a Mattermost Playbook Run using the "Security 
Vulnerability Playbook" and follow the process steps documented to continue with the Triage step. The playbook links to

> the Jira ticket created above, ensures the right process steps are completed, and informs relevant parties.

Reviewer: The Playbook should link to the bug ticket, not the vulnerability ticket, since the bug ticket is where most of the cross-team coordination happens and so that's the one that needs to be easily accessible.

> ###  Triage

Reviewer: IMO, in most cases triage should happen before running kicking off the playbook run. Only for complex / difficult-to-repro cases the playbook may be necessary for triage.

If the vulnerability is not confirmed valid or considered a duplicate, the security engineer must notify the reporter 
about the decision. If we don’t consider it a vulnerability, the justification for our decision should be included. The 
ticket should afterwards be closed either as duplicate or as invalid.

If the report is deemed a security improvement (but not vulnerability), a development ticket should be created and the 
security vulnerability ticket closed with the development ticket added as a reference. A report is considered a security 
improvement when it is not exploitable without the presence of any additional vulnerabilities or only increases the 
general security posture.

###  Fix Development
The Product Security Team must open a bug ticket and assign it to the corresponding feature team for the vulnerability 
found. The lead engineer/product manager of the feature team and the release manager should be notified with an 
“@”-mention about the security vulnerability in the Product Security channel and about its severity. The lead engineer 
must acknowledge the report, create a linked development ticket and assign it to an engineer. The lead engineer should 

> work together with the product manager on the prioritization of the reported security vulnerability.

Reviewer: With the playbook-based workflow this is no longer accurate.
